### PR TITLE
[FLINK-9349][Kafka Connector] KafkaConnector Exception while fetching from multiple kafka topics

### PR DESCRIPTION
**Pull request by snuyanzin**
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](http://flink.apache.org/contribute-code.html#best-practices).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.

**(The sections below can be removed for hotfixes of typos)**

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

fix synchronization issue
+ test to verify it

## Verifying this change

This change added tests and can be verified as follows:
via org.apache.flink.streaming.connectors.kafka.internal.Flink9349Test#testConcurrentPartitionsDiscoveryAndLoopFetching

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: ( no )
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? ( no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java [hunk 1]
```diff
@@ -507,7 +507,7 @@ private void updateMinPunctuatedWatermark(Watermark nextWatermark) {
 			SerializedValue<AssignerWithPunctuatedWatermarks<T>> watermarksPunctuated,
 			ClassLoader userCodeClassLoader) throws IOException, ClassNotFoundException {
 
-		List<KafkaTopicPartitionState<KPH>> partitionStates = new LinkedList<>();
+		List<KafkaTopicPartitionState<KPH>> partitionStates = new CopyOnWriteArrayList<>();
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: Would be nice to have a comment on why we need to use a `CopyOnWriteArrayList`

File: flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Flink9349Test.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+import static org.powermock.api.mockito.PowerMockito.whenNew;
+
+/**
+ * Unit tests for the {@link Flink9349Test}.
+ */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(KafkaConsumerThread.class)
+public class Flink9349Test {
+	@Test
+	public void testConcurrentPartitionsDiscoveryAndLoopFetching() throws Exception {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Flink9349Test.java]: I think we should have a similar test, but move it to `Kafka09FetcherTest`.

File: flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Flink9349Test.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+				}
+			});
+		}
+
+		service.awaitTermination(1L, TimeUnit.SECONDS);
+
+		// wait until the fetcher has reached the method of interest
+		sync.await();
+
+		// ----- trigger the offset commit -----
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Flink9349Test.java]: I don't think this is required for the scope of interest of this test, is it?

File: flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Flink9349Test.java (hunk 2 above)
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Flink9349Test.java]: Likewise, Kafka 08 / 09 / 010 / 011 should all have this test coverage.

**Comment by tzulitai** [comment]
Thanks for the PR @snuyanzin!
I had some comments, please let me know what you think.

Also, some general contribution tips:
1. I would suggest the title of the PR to be something along the lines of "[FLINK-9349] [kafka] Fix ConcurrentModificationException when add discovered partitions". That directly makes it clear what exactly is being fixed.
2. The message of the first commit of the PR should also be appropriately set to be similar to the title (most of the time if it is a 1-commit PR, the title of the PR and the commit message can be identical).

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java (hunk 1 above)
Comment (snuyanzin) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: Yes you are right. A question: is it allowed to specify a link on the issue comment where it was decided to use CopyOnWriteArrayList? Or is it better to have explanation in a comment only?

Comment (tedyu) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: explanation can be made with a comment.
No need to link to issue comment.

File: flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Flink9349Test.java (hunk 3 above)
Comment (snuyanzin) [reply flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/Flink9349Test.java]: thank you 
you are right I removed it and some other useless stuff from this test

**Comment by snuyanzin** [comment]
@tzulitai, @tedyu thnk you for your review, comments and contribution tips
I did updates which includes moving test into AbstractFetcherTest and making it kafka connector version independent

Could you please help me a bit?
Suddenly the travis build failed on YARNSessionCapacitySchedulerITCase (only on flink travis, on my fork it passed several times).  It does not look like result of changes as there is nothing related to yarn. Anyway I tried to investigate it. I found several similar issues on jira however they are closed. 

Also I downloaded logs mentioned in failed travis job 

> Uploading to transfer.sh
https://transfer.sh/JspTz/24547.10.tar.gz

based on them it looks like there was a connectivity issue with one of the ApplicationMaster 
as log yarn-tests/container_1526608500321_0007_01_000001/job-manager.log is full of
> 2018-05-18 01:56:49,448 WARN  akka.remote.transport.netty.NettyTransport                    - Remote connection to [null] failed with java.net.ConnectException: Connection refused: travis-job-2a2afdc5-7bf8-4597-946e-16551a5ebbc4/127.0.1.1:43980
2018-05-18 01:56:49,449 WARN  akka.remote.ReliableDeliverySupervisor                        - Association with remote system [akka.tcp://flink@travis-job-2a2afdc5-7bf8-4597-946e-16551a5ebbc4:43980] has failed, address is now gated for [50] ms. Reason: [Association failed with [akka.tcp://flink@travis-job-2a2afdc5-7bf8-4597-946e-16551a5ebbc4:43980]] Caused by: [Connection refused: travis-job-2a2afdc5-7bf8-4597-946e-16551a5ebbc4/127.0.1.1:43980]

very strange thing 

> Remote connection to [null]

**Comment by tzulitai** [comment]
@snuyanzin the failing `YARNSessionCapacitySchedulerITCase` is known to be bit flaky, so you can safely ignore that for now. I'll take another look at your changes soon. Thanks!

File: flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcherTest.java [hunk 4]
```diff
@@ -390,6 +398,102 @@ public void testPeriodicWatermarksWithNoSubscribedPartitionsShouldYieldNoWaterma
...
+		int fetchTasks = 5;
+		final CountDownLatch latch = new CountDownLatch(fetchTasks);
+		ExecutorService service = Executors.newFixedThreadPool(fetchTasks + 1);
+
+		service.submit(new Thread("fetcher runner") {
+			@Override
+			public void run() {
+				try {
+					latch.await();
+					fetcher.runFetchLoop();
```
[the remark this answers is not in the document]
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcherTest.java]: So, IMO, the test should look something like this:

```
		final OneShotLatch fetchLoopWaitLatch = new OneShotLatch();
		final OneShotLatch stateIterationBlockLatch = new OneShotLatch();

		final TestFetcher<String> fetcher = new TestFetcher<>(
			sourceContext,
			partitionsWithInitialOffsets,
			null, /* periodic assigner */
			null, /* punctuated assigner */
			new TestProcessingTimeService(),
			10,
			fetchLoopWaitLatch,
			stateIterationBlockLatch);

		// ----- run the fetcher -----

		final CheckedThread checkedThread = new CheckedThread() {
			@Override
			public void go() throws Exception {
				fetcher.runFetchLoop();
			}
		};
		checkedThread.start();

		// wait until state iteration begins before adding discovered partitions
		fetchLoopWaitLatch.await();
		fetcher.addDiscoveredPartitions(Collections.singletonList(testPartition));

		stateIterationBlockLatch.trigger();
		checkedThread.sync();
```

[the remark this answers is not in the document]
Comment (tzulitai) [reply flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcherTest.java]: The final `checkedThread.sync()` would always fail with the `ConcurrentModificationException` if the test is designed like this.

**Comment by snuyanzin** [comment]
@tzulitai thank you for your review and comments
based on your comments I have a question. Could you please clarify it?

You mentioned Flink's `OneShotLatch ` and `CheckedThread ` at the same time in some Kafka connector's tests used `AtomicReference`, `Thread` and etc. (I used one of them as an example while writing my version of the test). Just to be on the sage am I right that `OneShotLatch ` and `CheckedThread ` in tests are more preferable or are there some rules/limitations/whatever?

**Comment by tzulitai** [comment]
Using `CheckedThread` is more preferable, as it simplifies some of the test code.
But yes, the utility was introduced at a later point in time in Flink, so some parts of the test code might still be using `Thread`s and `AtomicReference`s.

**Pull request closed without merging**